airbyte-cdk/bulk/toolkits/load-avro/src/main/kotlin/io/airbyte/cdk/load/file/avro/AvroReader.kt

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.cdk.load.file.avro
+
+import java.io.Closeable
+import java.io.InputStream
+import kotlin.io.path.outputStream
+import org.apache.avro.Schema
+import org.apache.avro.file.DataFileReader
+import org.apache.avro.generic.GenericDatumReader
+import org.apache.avro.generic.GenericRecord
+
+class AvroReader(
+    private val dataFileReader: DataFileReader<GenericRecord>,
+    private val tmpFile: java.io.File
+) : Closeable {
+    private fun read(): GenericRecord? {
+        return if (dataFileReader.hasNext()) {
+            dataFileReader.next()
+        } else {
+            null
+        }
+    }
+
+    fun recordSequence(): Sequence<GenericRecord> {
+        return generateSequence { read() }
+    }
+
+    override fun close() {
+        dataFileReader.close()
+        tmpFile.delete()
+    }
+}
+
+fun InputStream.toAvroReader(avroSchema: Schema): AvroReader {
+    val reader = GenericDatumReader<GenericRecord>(avroSchema)
+    val tmpFile =
+        kotlin.io.path.createTempFile(
+            prefix = "${avroSchema.namespace}.${avroSchema.name}",
+            suffix = ".avro"
+        )
+    tmpFile.outputStream().use { outputStream -> this.copyTo(outputStream) }
+    val file = tmpFile.toFile()
+    val dataFileReader = DataFileReader(file, reader)
```

Comment:
does this class work better? https://avro.apache.org/docs/1.8.0/api/java/org/apache/avro/file/DataFileStream.html#DataFileStream(java.io.InputStream,%20org.apache.avro.io.DatumReader)

(it accepts an inputstream directly, but presents a different interface)*

Reply:
No, I tried that first. DatumWriter/Reader just writes binary avro without file headers, etc. DataFileReader/Writer read/write in file *format*. It's just that FileWriter has a dumb output stream interface, but FileReader requires a seekable stream.